### PR TITLE
Skip printing the backtrace for RecvError/SendError

### PR DESCRIPTION
**Pull request by Manishearth**
We currently get tons of useless backtraces clogging up the output when we have a panic cascade. This adds a handler that outputs a single line when a thread panics due to a sender or receiver hanging up, since this is almost always due to a panic cascade.

We could add a commandline arg that gets us back the old behavior, though I'm not sure if this is necessary.

r? @asajeffrey

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10329)

### CODE REVIEW

File: components/util/thread.rs [hunk 1]
```diff
@@ -10,11 +10,38 @@ use std::thread;
...
     let builder = thread::Builder::new().name(name);
-    builder.spawn(f).unwrap();
+
+    let f_with_handler = move || {
+        let hook = take_handler();
+
+        let new_handler = move |info: &PanicInfo| {
+            let payload = info.payload();
+            if let Some(s) = payload.downcast_ref::<String>() {
+                if s.contains("SendError") {
```
Comment (Manishearth) [inline components/util/thread.rs]: We could perhaps improve this check with a better regex match.

**Comment by Manishearth** [comment]
We still get panic cascades with IPC, however I'm not sure if that's due to https://github.com/servo/servo/issues/10260 or is a general thing/

File: components/util/thread.rs [hunk 2]
```diff
@@ -5,16 +5,41 @@
...
 use serde::Serialize;
 use std::borrow::ToOwned;
+use std::io::{Write, stderr};
+use std::panic::{PanicInfo, take_handler, set_handler};
 use std::sync::mpsc::Sender;
 use std::thread;
 use std::thread::Builder;
 use thread_state;
-
+#[allow(unused_must_use)]
```
Comment (Manishearth) [inline components/util/thread.rs]: I skip the result of `write!`. I don't want to panic within a panic handler, and if stderr is broken I don't see much point of trying to output an error either.

Comment (eddyb) [inline components/util/thread.rs]: `let _ = ...;`

Comment (Manishearth) [inline components/util/thread.rs]: done

**Comment by jdm** [comment]
My fear is that this will suppress useful backtraces for errors that are not part of a panic cascade that happen to involve sending or receiving on an invalid sender/receiver.

**Comment by Manishearth** [comment]
That's the only time you can get SendError/RecvError

**Comment by asajeffrey** [comment]
Can't you get a SendError/RecvError if the other half of the channel has been reclaimed? So this would eat the stack trace even in cases where there's not a panic cascade.

In production use, there's a request (#10334) to send an event to the chrome, which can determine how best to present panics to the user. We might want to rely on that for the UX of panics, rather than trying to do a regex-based edit.

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10329#-:-KEH-FBSDnrgADYx1z0H:-1314421906)*

**Comment by Manishearth** [comment]
I could add a cmdline option to get it back.

I'm not sure how #10334 can be handled. By the time you reach a panic handler, that thread is effectively gone already -- the relevant machinery to report that error is already missing. `recover()` might be useful here, not sure.

**Comment by Manishearth** [comment]
I added a commandline option.

As far as #10334, one way to handle this might be to have a global privileged "parent window". Windows corresponding to browser.html and those using that API have this. On Servo's side we store a global once-initialized static channel to its script task, and in the panic handler _if_ we find such a channel _and_ sending messages to the window works (use `try_send()` instead of `send()`), send a message to the script task with panic details. This triggers an event on the bhtml `Window`.

The alternative is to have each `iframe mozbrowser` get a special custom panic handler that bubbles it up. However, panics may not always be in the per-iframe threads, and if something else panics then there's not much we can do about it.

**Comment by asajeffrey** [comment]
IRC conversation at http://logs.glob.uno/?c=mozilla%23servo#c399939

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10329#-:-KEYtgSQNMxLX_BxHpnK:1349752157)*

**Comment by Manishearth** [comment]
@bors-servo retry

```
Tests with unexpected results:
  ▶ TIMEOUT [expected OK] /cssom-view/scrollingElement.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  │ ERROR:js::rust: Error at http://web-platform.test:8000/cssom-view/scrollingElement.html:14:4: URL.createObjectURL is not a function
  │ 
  │ ERROR:js::rust: Error at http://web-platform.test:8000/cssom-view/scrollingElement.html:11:6: URL.createObjectURL is not a function
  └ 
```

this sounds like a problem with the infra, let's see if it's intermittent

**Comment by KiChjang** [comment]
@Manishearth That's #10398.

**Comment by KiChjang** [comment]
@bors-servo rery
* #10398

**Comment by KiChjang** [comment]
@bors-servo retry
* #10398

**Comment by Manishearth** [comment]
Huh, looks like github search can't find that if I search for XFree

**Comment by KiChjang** [comment]
@bors-servo retry
* #10398

**Pull request merged**